src/ray/core_worker/test/actor_task_submitter_test.cc

```diff
@@ -654,8 +635,28 @@ TEST_P(ActorTaskSubmitterTest, TestActorRestartFailInflightTasks) {
   const auto death_cause = CreateMockDeathCause();
   submitter_.DisconnectActor(
       actor_id, 1, /*dead=*/false, death_cause, /*is_restartable=*/true);
+  ASSERT_EQ(submitter_.NumInflightTasks(actor_id), 0);
+  // RPC callbacks are delayed.
+  ASSERT_EQ(worker_client_->callbacks.size(), 2);
 
-  // The task replies are now received. Since the tasks are already failed, they will not
+  // Task2 and task3 are retried.
+  task2.GetMutableMessage().set_attempt_number(task2.AttemptNumber() + 1);
+  task2.GetMutableMessage().mutable_actor_task_spec()->set_actor_counter(3);
+  task3.GetMutableMessage().set_attempt_number(task3.AttemptNumber() + 1);
+  task3.GetMutableMessage().mutable_actor_task_spec()->set_actor_counter(4);
+  ASSERT_TRUE(CheckSubmitTask(task2));
+  ASSERT_TRUE(CheckSubmitTask(task3));
+
+  // Actor is restarted.
+  rpc::Address actor_addr2;
+  actor_addr2.set_worker_id(WorkerID::FromRandom().Binary());
+  actor_addr2.set_port(1);
+  submitter_.ConnectActor(actor_id, actor_addr2, 1);
+  ASSERT_EQ(worker_client_->callbacks.size(), 4);
+  ASSERT_EQ(submitter_.NumInflightTasks(actor_id), 2);
+
+  // The task replies of the first attempt are now received.
+  // Since the first attempts are already failed, they will not
```

Comment:
we should also add another case that tests when these RPC responses aren't received until after the second attempts have completed